File: src/utils/misc/source/logging.cpp
/* 
 *  iviLINK SDK, version 0.9 (for preview only)                                      
 *    http://www.ivilink.net                                                         
 *  Cross Platform Application Communication Stack for In-Vehicle Applications       
 *                                                                                   
 *  Copyright (C) 2012, Luxoft Professional Corp., member of IBS group               
 *                                                                                   
 *  This library is free software; you can redistribute it and/or                    
 *  modify it under the terms of the GNU Lesser General Public                       
 *  License as published by the Free Software Foundation; version 2.1.               
 *                                                                                   
 *  This library is distributed in the hope that it will be useful,                  
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of                   
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU                
 *  Lesser General Public License for more details.                                  
 *                                                                                   
 *  You should have received a copy of the GNU Lesser General Public                 
 *  License along with this library; if not, write to the Free Software              
 *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA   
 * 
 * 
 * 
 * 
 */





#include <assert.h>
#include <cstdarg>
#include <ctime>
#include <cstdio>
#include <pthread.h>
#include <stdint.h>

#include "utils/misc/include/Types.hpp"
#include "utils/misc/include/logging.hpp"
#include <unistd.h>

#if defined(ANDROID)
   #include <sys/syscall.h>
   #include <android/log.h>
#else
   #include <syscall.h>
#endif

static const char gDefaultPrefix[] = "";
static char const* gpPrefix = gDefaultPrefix;

void Log_SetPrefix(char const* const pPrefix)
{
   if (pPrefix)
      gpPrefix = pPrefix;
   else
      gpPrefix = gDefaultPrefix;
}

#if defined(ANDROID)
void _Log(const char* tag, const char* format, ...)
#else
void _Log(const char* format, ...)
#endif
{
   time_t curTime;
   tm tmCurTime;
   const size_t bufSize = 1024;
   char buf[bufSize] = "";
   size_t size = 0;

   size += snprintf(buf + size, bufSize - size, "%s", gpPrefix);

#if !defined(ANDROID)
   struct timespec acttime;

   clock_gettime(CLOCK_REALTIME, &acttime);

   if (time(&curTime) != (time_t) (-1))
   {
      if (localtime_r(&curTime, &tmCurTime) != NULL)
      {
         size += strftime(buf + size, bufSize - size, "[%T", &tmCurTime);
         //printf("[%s] ", buf);
      }
   }
   size += snprintf(buf + size, bufSize - size, ":%03ld]",
            acttime.tv_nsec / 1000000L);
#endif

#if defined(ANDROID)
   size += snprintf(buf + size, bufSize - size, "{TID %lu}", gettid());
#else
   size += snprintf(buf + size, bufSize - size, "{TID %lu:%i}", pthread_self(), getpid());
#endif // ANDROID

   va_list ap;
   va_start(ap, format);
   vsnprintf(buf + size, bufSize - size, format, ap);
   va_end(ap);

#if defined(ANDROID)
   __android_log_write(ANDROID_LOG_DEBUG, tag, buf);
#else
   fprintf(stderr, "%s\n", buf);
   fflush(stderr);
#endif
}

#if defined(ANDROID)
void _hex_print(const char* tag, const void* arr, unsigned int size)
#else
void _hex_print(const void* arr, unsigned int size)
#endif
{
   const UInt8* a = (const UInt8*) arr;
   const size_t bufSize = 1024;
   //assert(size * 8 <= bufSize);
   if (size * 8 >= bufSize)
      return;
   char buf[bufSize] = "";

   size_t cnt;
   size_t buf_c = 0;
   size_t ln_beg;

   for (ln_beg = cnt = 0; cnt < size; cnt += 1)
   {
      char b = (a[cnt] & 0xf0) >> 4;
      buf[buf_c++] = b > 9 ? 'a' + b - 10 : '0' + b;
      b = a[cnt] & 0x0f;
      buf[buf_c++] = b > 9 ? 'a' + b - 10 : '0' + b;

      buf[buf_c++] = ' ';
      if ((cnt & 0x03) == 0x03)
         buf[buf_c++] = ' ';

      if ((cnt & 0x07) == 0x07 || cnt + 1 == size)
      {
         buf[buf_c++] = ' ';
         for (; ln_beg <= cnt; ln_beg++)
         {
            buf[buf_c++] = a[ln_beg] < 32 ? '.' : a[ln_beg];
         }
         ln_beg = cnt + 1;

         buf[buf_c++] = '\n';
      }
   }
   buf[buf_c++] = 0;

#if defined(ANDROID)
   __android_log_write(ANDROID_LOG_DEBUG, tag, buf);
#else
   fwrite(buf, 1, buf_c, stderr);
   fflush(stderr);
#endif
}

File: iviLink/include_android/java/Utils/src/com/luxoft/ivilink/sdk/android/lib/utils/log/Logging.java
/* 
 * iviLINK SDK, version 1.2
 * http://www.ivilink.net
 * Cross Platform Application Communication Stack for In-Vehicle Applications
 * 
 * Copyright (C) 2012-2013, Luxoft Professional Corp., member of IBS group
 * 
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; version 2.1.
 * 
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
 * 
 */ 
package com.luxoft.ivilink.sdk.android.lib.utils.log;

import java.nio.ByteBuffer;
import java.util.Formatter;

import android.util.Log;

/**
 * Provides some helpful functions for logging.
 */
public class Logging {

	private static final int CALLER_METHOD_STACK_INDEX;

	static {
		int i = 0;
		for (StackTraceElement ste : Thread.currentThread().getStackTrace()) {
			i++;
			if (ste.getClassName().equals(Logging.class.getName())) {
				break;
			}
		}
		Log.i(Logging.class.getCanonicalName(), "CALLER_METHOD_STACK_INDEX = " + i);
		CALLER_METHOD_STACK_INDEX = i;
	}

	/**
	 * 
	 * @return Name of the method in which this one was called
	 */
	public static String getCurrentMethodName() {
		StackTraceElement[] ste = Thread.currentThread().getStackTrace();
		return "CURRENT METHOD: " + ste[CALLER_METHOD_STACK_INDEX].getMethodName() + "(...)";
	}

	/**
	 * 
	 * @param args
	 *            List of arguments of the caller method should be placed here
	 * @return Name of the method in which this one was called with list of
	 *         arguments in human-readable format
	 */
	public static String getCurrentMethodName(Object... args) {
		StackTraceElement[] ste = Thread.currentThread().getStackTrace();
		StringBuffer buffer = new StringBuffer();
		if (args == null) {
			buffer.append("null");
		} else {
			for (int i = 0; i < args.length; i++) {
				if (args[i] != null) {
					buffer.append(args[i].toString());
				} else {
					buffer.append("null");
				}
				buffer.append(i == args.length - 1 ? "" : ", ");
			}
		}
		return "CURRENT METHOD: " + ste[CALLER_METHOD_STACK_INDEX].getMethodName() + "("
				+ buffer.toString() + ")";
	}

	/**
	 * Converts byte array into a hex String
	 * 
	 * @param bytes
	 *            Bytes to convert
	 * @return Converted String
	 */
	public static String bytesToHexString(byte[] bytes) {
		StringBuilder sb = new StringBuilder(bytes.length * 3);

		Formatter formatter = new Formatter(sb);
		for (byte b : bytes) {
			formatter.format(" %02x", b);
		}
		formatter.close();
		return sb.toString();
	}

	  /**
	   * Converts contents of a ByteBuffer into a hex String
	   * 
	   * @param bytes
	   *            Bytes to convert
	   * @return Converted String
	   */
	  public static String bytesToHexString(ByteBuffer bytes) {
	      return bytesToHexString(bytes.array());
	  }
}
